pkg/server/container_execsync.go

```diff
@@ -17,15 +17,134 @@ limitations under the License.
+	glog.V(2).Infof("ExecSync for %q with command %+v and timeout %d (s)", r.GetContainerId(), r.GetCmd(), r.GetTimeout())
+	defer func() {
+		if retErr == nil {
+			glog.V(2).Infof("ExecSync for %q returns with exit code %d", r.GetContainerId(), retRes.GetExitCode())
+			glog.V(4).Infof("ExecSync for %q outputs - stdout: %q, stderr: %q", r.GetContainerId(),
+				retRes.GetStdout(), retRes.GetStderr())
+		}
+	}()
+
+	// Get container config from container store.
+	meta, err := c.containerStore.Get(r.GetContainerId())
+	if err != nil {
+		return nil, fmt.Errorf("an error occurred when try to find container %q: %v", r.GetContainerId(), err)
+	}
+	id := r.GetContainerId()
+
+	if meta.State() != runtime.ContainerState_CONTAINER_RUNNING {
+		return nil, fmt.Errorf("container %q is in %s state", id, criContainerStateToString(meta.State()))
+	}
+
+	// TODO(random-liu): Replace the following logic with containerd client and add unit test.
+	// Prepare streaming pipes.
+	execDir, err := ioutil.TempDir(getContainerRootDir(c.rootDir, id), "exec")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create exec streaming directory: %v", err)
+	}
+	defer func() {
+		if err = c.os.RemoveAll(execDir); err != nil {
+			glog.Errorf("Failed to remove exec streaming directory %q: %v", execDir, err)
+		}
+	}()
+	_, stdout, stderr := getStreamingPipes(execDir)
+	_, stdoutPipe, stderrPipe, err := c.prepareStreamingPipes(ctx, "", stdout, stderr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare streaming pipes: %v", err)
+	}
+	defer stdoutPipe.Close()
+	defer stderrPipe.Close()
+
+	// Start redirecting exec output.
+	stdoutBuf, stderrBuf := new(bytes.Buffer), new(bytes.Buffer)
+	go io.Copy(stdoutBuf, stdoutPipe) // nolint: errcheck
+	go io.Copy(stderrBuf, stderrPipe) // nolint: errcheck
+
+	// Get containerd event client first, so that we won't miss any events.
+	// TODO(random-liu): Handle this in event handler. Create an events client for
+	// each exec introduces unnecessary overhead.
+	cancellable, cancel := context.WithCancel(ctx)
+	events, err := c.taskService.Events(cancellable, &execution.EventsRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get containerd event: %v", err)
+	}
+
+	spec := &meta.Spec.Process
+	spec.Args = r.GetCmd()
+	rawSpec, err := json.Marshal(spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal oci spec %+v: %v", spec, err)
+	}
+
+	resp, err := c.taskService.Exec(ctx, &execution.ExecRequest{
+		ContainerID: id,
+		Terminal:    false,
+		Stdout:      stdout,
+		Stderr:      stderr,
+		Spec: &prototypes.Any{
+			TypeUrl: runtimespec.Version,
+			Value:   rawSpec,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to exec in container %q: %v", id, err)
+	}
+	exitCode, err := waitContainerExec(cancel, events, id, resp.Pid, r.GetTimeout())
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for exec in container %q to finish: %v", id, err)
```

Comment:
If containerd dies during waiting, the exec process may still be running, so I think we should return error instead of 255 exit code without error.

I still keep the unknownExitCode, because:
1) We may still need it when adding the recovery logic after daemon restart.
2) Just in case we use the 0 exit code by mistake.